server/src/main/java/io/unitycatalog/server/persist/UserRepository.java

```diff
@@ -0,0 +1,191 @@
+package io.unitycatalog.server.persist;
+
+import io.unitycatalog.control.model.CreateUser;
+import io.unitycatalog.control.model.UpdateUser;
+import io.unitycatalog.control.model.User;
+import io.unitycatalog.server.exception.BaseException;
+import io.unitycatalog.server.exception.ErrorCode;
+import io.unitycatalog.server.persist.dao.UserDAO;
+import io.unitycatalog.server.persist.utils.HibernateUtils;
+import io.unitycatalog.server.persist.utils.PagedListingHelper;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UserRepository {
+  private static final UserRepository INSTANCE = new UserRepository();
+  private static final Logger LOGGER = LoggerFactory.getLogger(UserRepository.class);
+  private static final SessionFactory SESSION_FACTORY = HibernateUtils.getSessionFactory();
+  private static final PagedListingHelper<UserDAO> LISTING_HELPER =
+      new PagedListingHelper<>(UserDAO.class);
+
+  private UserRepository() {}
+
+  public static UserRepository getInstance() {
+    return INSTANCE;
+  }
+
+  public User createUser(CreateUser createUser) {
+    User user =
+        new User()
+            .id(UUID.randomUUID().toString())
+            .name(createUser.getName())
+            .email(createUser.getEmail())
+            .externalId(createUser.getExternalId())
+            .state(User.StateEnum.ENABLED)
+            .createdAt(System.currentTimeMillis());
+
+    try (Session session = SESSION_FACTORY.openSession()) {
+      Transaction tx = session.beginTransaction();
+      try {
+        if (getUserByEmail(session, user.getEmail()) != null) {
+          throw new BaseException(
+              ErrorCode.ALREADY_EXISTS, "User already exists: " + user.getEmail());
+        }
+        session.persist(UserDAO.from(user));
+        tx.commit();
+        return user;
+      } catch (Exception e) {
+        tx.rollback();
+        throw e;
+      }
+    }
+  }
+
+  public List<User> listUsers() {
+    return listUsers(false);
+  }
+
+  public List<User> listUsers(boolean includeDisabled) {
+    try (Session session = SESSION_FACTORY.openSession()) {
+      session.setDefaultReadOnly(true);
+      Transaction tx = session.beginTransaction();
+      try {
+        List<UserDAO> userDAOs =
```

Comment:
This is unpaginated (even though the API itself has a count), which it can be expensive on the DB. For example, if there are 100K total users and the count is 10, this would fetch all the 100K users from the DB and throw away all except 10.

Understandably, the `filter` parameter in the API complicates pagination, but there are ways to implement it that is closer to O(count) and not O(total number of users)  (e.g., by paging it from the DB, with potentially multiple small DB queries).

This doesn't need to be fixed for now, but can you create a github issue to fix it in the future?

Reply:
Actually, the `PagedListingHelper` defaults to a max count of 100 if `maxResultsOpt` is not present. So, I think the `listUsers` API can return fewer users than requested and isn't correct.

Reply:
Most of the Databricks paginated APIs say they can return fewer users than requested, so its consistent with that. Even so, this is a little trickier for SCIM since its "pagination" works differently. We'll have to paginate internally until we get to the requested SCIM `startIndex` and then gather results until we get enough elements.

Reply:
Ok this has been reworked to get a clock from `PagedListingHelper` and skip and return max as appropriate.

Reply:
Done.